gslib/tests/test_psc.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for private service connect custom endpoints."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import division
+from __future__ import unicode_literals
+
+from boto import config
+
+from gslib.gcs_json_api import DEFAULT_HOST
+from gslib.tests import testcase
+from gslib.tests.testcase import integration_testcase
+from gslib.tests.util import ObjectToURI
+from gslib.tests.util import SetBotoConfigForTest
+from gslib.tests.util import unittest
+
+# Get full output from stdout, which might not be piped in Python 3.
+PYTHON_UNBUFFERED_ENV_VAR = {'PYTHONUNBUFFERED': '1'}
+
+
+class TestPsc(testcase.GsUtilIntegrationTestCase):
+  """Integration tests for PSC custom endpoints."""
+
+  @integration_testcase.SkipForXML('JSON test.')
+  @integration_testcase.SkipForS3('Custom endpoints not available for S3.')
+  def test_persists_custom_endpoint_through_json_sliced_download(self):
+    gs_host = config.get('Credentials', 'gs_json_host', DEFAULT_HOST)
+    if gs_host == DEFAULT_HOST:
+      # Captures case where user may have manually set gs_json_host to the
```

Comment:
Even if the user does not set anything, it will enter this case, correct? Since gs_host will get set to DEFAULT_HOST? I think this check was added  to avoid running this test for non-kokoro env? Maybe specifying that reason would be more helpful.

Reply:
No, this isn't to stop the test from running outside Kokoro. It's fine to run the test outside Kokoro as long as the user has a working custom endpoint. The reason is in the code comment: This test will fail if the user sets a "custom" endpoint that is equal to the default endpoint. I.e. The user set their `.boto` to have `gs_json_host=storage.googleapis.com`.

Reply:
Isn't default=storage.googleapis.com anyway?

so if nothing is set, gs_host will be "storage.googleapis.com" and this condition will be true?

Reply:
Discussed offline. Summary:
We don't want this test to run if there isn't a custom endpoint. There is not a custom endpoint if:
1) gs_host and gs_json_host are not set.
2) The higher priority of the above is set, but the user set it to the default endpoint.